File: runtime/org.eclipse.etrice.runtime.cpp/src/common/messaging/MessageDispatcher.cpp
/*******************************************************************************
 * Copyright (c) 2012 Draeger Medical GmbH (http://www.draeger.com).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * CONTRIBUTORS:
 * 		Peter Karlitschek (initial contribution)
 *
 *******************************************************************************/

#include "common/messaging/Message.h"
#include "common/messaging/MessageDispatcher.h"
#include "common/messaging/IMessageService.h"

namespace etRuntime {

MessageDispatcher::MessageDispatcher(IRTObject* parent, const Address& addr, const String& name) :
		RTObject(parent, name),
		m_local_map(),
		m_freeAdresses(),
		m_pollingMessageReceiver(),
		m_address(addr),
		m_nextFreeObjId(addr.m_objectID + 1) {
	m_local_map[addr.m_objectID] = this;
}

Address MessageDispatcher::getFreeAddress() {
	if (m_freeAdresses.empty()) {
		return Address(getAddress().m_nodeID, getAddress().m_threadID, m_nextFreeObjId++);
	}
	else {
		Address address = m_freeAdresses.back();
		m_freeAdresses.pop_back();

		return address;
	}
}
void MessageDispatcher::freeAddress(const Address& addr) {
	m_freeAdresses.push_back(addr);
}
void MessageDispatcher::addMessageReceiver(IMessageReceiver& receiver) {
	if (!receiver.getAddress().isValid())
		return;

	if (receiver.getAddress().m_nodeID == m_address.m_nodeID
			&& receiver.getAddress().m_threadID == m_address.m_threadID) {

		m_local_map[receiver.getAddress().m_objectID] = &receiver;
	}

}

void MessageDispatcher::removeMessageReceiver(IMessageReceiver& receiver) {
	if (!receiver.getAddress().isValid())
		return;

	m_local_map.erase(receiver.getAddress().m_objectID);
}

void MessageDispatcher::addPollingMessageReceiver(IMessageReceiver& receiver) {
	m_pollingMessageReceiver.insert(&receiver);
}
void MessageDispatcher::removePollingMessageReceiver(IMessageReceiver& receiver) {
	m_pollingMessageReceiver.erase(&receiver);
}

void MessageDispatcher::receive(const Message* msg) {

	IMessageReceiver* receiver = 0;
	if (msg->getAddress().m_nodeID == m_address.m_nodeID && msg->getAddress().m_threadID == m_address.m_threadID) {
		Map<int, IMessageReceiver*>::iterator it = m_local_map.find(msg->getAddress().m_objectID);
		if (it != m_local_map.end()) {
			receiver = it->second;
		}
	}
	if (receiver == this) {
		for (Set<IMessageReceiver*>::iterator it = m_pollingMessageReceiver.begin();
				it != m_pollingMessageReceiver.end(); ++it) {
			(*it)->receive(msg);
		}
	} else if (receiver != 0) {
		receiver->receive(msg);
		// TODO: error handling for not found addresses
	}

	msg->~Message();
	dynamic_cast<IMessageService*>(getParent())->returnMessageBuffer(msg);
}

String MessageDispatcher::toString() const {
	return getName() + " " + getAddress().toID().c_str();
}

} /* namespace etRuntime */

File: runtime/org.eclipse.etrice.runtime.java/src/org/eclipse/etrice/runtime/java/messaging/MessageDispatcher.java
/*******************************************************************************
 * Copyright (c) 2010 protos software gmbh (http://www.protos.de).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *******************************************************************************/

package org.eclipse.etrice.runtime.java.messaging;

import java.util.LinkedList;
import java.util.HashMap;

/**
 * The message dispatcher class used by the MessageService.
 * 
 * @author Thomas Schuetz (initial contribution)
 * @author Henrik Rentz-Reichert (address management)
 *
 */
public class MessageDispatcher extends RTObject implements IMessageReceiver {

	private HashMap<Number, IMessageReceiver> local_map = new HashMap<Number, IMessageReceiver>();
	private LinkedList<Address> freeAdresses = new LinkedList<Address>();
	
	private Address address = null;
	private int nextFreeObjId;

	public MessageDispatcher(IRTObject parent, Address addr, String name){
		super(parent, name);
		address = addr;
		nextFreeObjId = addr.objectID+1;
	}
	
	public Address getFreeAddress() {
		if (freeAdresses.isEmpty()) {
			return new Address(getAddress().nodeID, getAddress().threadID, nextFreeObjId++);
		}
		else
			return freeAdresses.remove();
	}
	
	public void freeAddress(Address addr) {
		freeAdresses.add(addr);
	}
	
	public void addMessageReceiver(IMessageReceiver receiver){
		if (receiver.getAddress()==null)
			return;
		
		if (receiver.getAddress().nodeID == address.nodeID
				&& receiver.getAddress().threadID == address.threadID) {
			local_map.put(receiver.getAddress().objectID, receiver);
		}
	}
	
	public void removeMessageReceiver(IMessageReceiver receiver){
		if (receiver.getAddress()==null)
			return;
		
		if (receiver.getAddress().nodeID == address.nodeID
				&& receiver.getAddress().threadID == address.threadID) {
			local_map.remove(receiver.getAddress().objectID);
		}
	}
	
	@Override
	public void receive(Message msg) {
		if (msg.getAddress().nodeID == address.nodeID
				&& msg.getAddress().threadID == address.threadID) {
			IMessageReceiver receiver = local_map.get(msg.getAddress().objectID);
			if(receiver!=null) {
				receiver.receive(msg);
			}
			else {
				// TODO: error handling for not found addresses
			}
		}
	}
	
	@Override
	public Address getAddress() {
		return address;
	}
}
